design-proposals/vm-memory-dump.md

```diff
@@ -0,0 +1,101 @@
+# Overview
+Proposal for taking VM memory dump for analysis purposes.
+
+## Motivation
+One of the methods of troubleshooting guest virtual machines is getting a memory dump and analyzing it. As this is a standard capability in most VM management systems, it is also required for kubevirt’s VMs.
+
+## Goals
+Have a mechanism to get a VM memory dump.
+
+## Non Goals
+Memory dump is not a restorable output and currently will not be a part of the VM snapshot.
+
+## User Stories
+As a Kubevirt user I would like to get a memory dump of a running VM so that I can later inspect its memory.
+
+## Repos
+[KubeVirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+To trigger a memory dump a virtctl memoryDump command will be added.
+The command will either use an existing PVC or create a new PVC on demand. This PVC will be bound to the VM and will appear in the VM volumes spec with a fitting status.
+During the memory dump process this PVC will be mounted to the virt-launcher pod. After the mounting the guest memory will be dumped to that pvc and eventually it will be unmounted from the virt launcher.
+The PVC will remain bound - in the VM spec. It is the users' responsibility to export/use the memory dump results before reusing the PVC for a new memory dump. As long as the PVC is bound as a memory dump "container" each memory dump command will overwrite the previous content.
+It will be possible to track the last memory dump in the VM status with a new memory dump status that will be added there.
+It will be possible to unbound the PVC and then do a memory dump to another PVC that will be bound instead.
+
+## API
+
+### Get memory dump
+#### trigger memory dump
+Run a virtctl command to get a memory dump. This new API will look as follows:
+
+`$ virtctl memory-dump my-vm --volume-name=memory-pvc`
```

Comment:
`--volume-name` optional if already in the volumes list or no?

Reply:
yes maybe, I'll see when implementing if it works out

Reply:
What is the purpose of including the memory dump pvc in the list of volumes?  I can understand it being there while a dump is in progress, but whey after?

Reply:
hmmm, the idea is that if it is in the vm spec it can be a part of the vmsnapshot since it is counted as part of the volumes. If you think we can just handle it specially in the vmsnapshot and your export doesnt also require it to be part of the vm volumes list then maybe in the vm spec volumes list we dont need to update it there at all and the memorydumpstatus is enough

Reply:
Okay this is fine, but the snapshot will not generate a dump (at least initially), right?

Reply:
no it will not (not initially at least)